compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -90,6 +93,79 @@ pub struct DefCollector {
     pub(crate) collected_traits_impls: TraitImplMap,
 }
 
+#[derive(Default)]
+pub struct CompilationErrors {
+    pub parser_errors: Vec<(ParserError, FileId)>,
+    pub def_errors: Vec<(DefCollectorErrorKind, FileId)>,
+    pub resolver_errors: Vec<(ResolverError, FileId)>,
+    pub type_errors: Vec<(TypeCheckError, FileId)>,
+}
```

Comment:
Instead of a struct containing each error type, why not have an enum of each error type as a variant? I don't think we'd need quite so many changes to support it in that case and we could eventually replace `CustomDiagnostic` with this enum.

Reply:
Expressed is as pseudo code, it this what you are suggesting ?
```
#[derive(Default)]
pub struct CompilationErrors {
    pub parser_errors: Vec<(ParserError, FileId)>,
    pub def_errors: Vec<(DefCollectorErrorKind, FileId)>,
    pub resolver_errors: Vec<(ResolverError, FileId)>,
    pub type_errors: Vec<(TypeCheckError, FileId)>,
}
```
To become something like this.  

```
pub enum CompilationError {
    ParseError(ParserError),
    DefinitionError(DefCollectorErrorKind),
    ResolveError(ResolverError),
    TypeError(TypeCheckError),
}

struct CompilationErrors {
    pub errors: Vec<(CompilationError, FileId)>,
}

impl CompilationErrors {
    fn has_errors(&self) -> bool {
        ....
    }
    fn has_parser_errors(&self) -> bool {
        ...
    }
}
```

This is actually doable and it seems like it's better design. I will implement it.

Reply:
Ah, I was hoping we wouldn't need the wrapper struct either but I suppose we do need the FileIds

Reply:
Please, review the new changes. Thanks